core/modules/replica/JobController.cc

```diff
@@ -0,0 +1,503 @@
+                                                        // be always used to wait for a specific interval
+
+                    unsigned int const wakeUpIvalMillisec =
+                        1000 * self->_serviceProvider->config()->jobSchedulerIvalSec();
+
+                    while (blockPost.wait(wakeUpIvalMillisec)) {
+
+                        // Initiate the stopping sequence if requested
+                        if (self->_stop) {
+
+                            // Cancel all outstanding jobs
+                            self->cancelAll();
+
+                            // Block here waiting before the controller will stop
+                            self->_controller->stop();
+
+                            // Quit the thread
+                            return;
+                        }
+
+                        // Check if there are jobs scheduled to run on the periodic basis
+                        self->runScheduled();
+                    }
+                }
+            )
+        );
+    }
+}
+
+bool JobController::isRunning() const {
+    return _thread.get() != nullptr;
+}
+
+void JobController::stop() {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::stop");
+
+    // IMPORTANT:
+    //
+    //   Never attempt running these operations within util::Lock lock(_mtx)
+    //   due to a possibile deadlock when asynchronous handlers will be
+    //   calling the thread-safe methods. A problem is that until they finish
+    //   in a clean way the thread will never finish, and the application will
+    //   hang on _thread->join().
+
+    // util::Lock lock(_mtx)  (disabled)
+
+    if (not isRunning()) return;
+
+    // Tell the thread to finish
+    _stop = true;
+
+    // Join with the thread before clearning up the pointer
+    _thread->join();
+    _thread.reset(nullptr);
+
+    _stop = false;
+}
+
+void JobController::join() {
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::join");
+    if (_thread) _thread->join();
+}
+
+FindAllJob::Ptr JobController::findAll(std::string const& databaseFamily,
+                                       bool saveReplicaInfo,
+                                       FindAllJob::CallbackType onFinish,
+                                       Job::Options const& options) {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::findAll");
+
+    util::Lock lock(_mtx, "JobController::findAll");
+
+    auto const self = shared_from_this();
+    auto const job =
+        FindAllJob::create(
+            databaseFamily,
+            saveReplicaInfo,
+            _controller,
+            std::string(),
+            [self] (FindAllJob::Ptr job) {
+                self->onFinish (job);
+            },
+            options
+        );
+
+    // Register the job (along with its callback) by its unique
+    // identifier in the local registry. Once it's complete it'll
+    // be automatically removed from the Registry.
+
+    _registry[job->id()] =
```

Comment:
there's no risk of a duplicate job id in the registry?

Reply:
No risk. Because JobController also acts as a job factory, and it only registers job objects created locally, and each job object has a unique identifier.